errorcodes/errorcodes.go

```diff
@@ -0,0 +1,241 @@
+
+var NoErrorReason = WithErrorReason(NoReason, 503, codes.Internal)
+
+type ResourceInfo struct {
+	ResourceType string
+	ResourceName string
+}
+
+func WithErrorReason(reason string, httpCode int, grpcStatusCode codes.Code) ErrorOption {
+	f := func(er *DaprError) {
+		er.reason = reason
+		er.grpcStatusCode = grpcStatusCode
+		er.httpCode = httpCode
+	}
+	return f
+}
+
+// call this function to apply option
+type ErrorOption func(*DaprError)
+
+type DaprError struct {
+	err            error
+	description    string
+	reason         string
+	httpCode       int
+	grpcStatusCode codes.Code
+	metadata       map[string]string
+	resourceInfo   *ResourceInfo
+}
+
+// EnableComponentErrorCode stores an indicator for
+// components to determine if the ErrorCodes Feature
+// is enable.
+func EnableComponentErrorCode(metadata map[string]string) {
+	if metadata != nil {
+		metadata[ErrorCodesFeatureMetadataKey] = "true"
+	}
+}
+
+func featureEnabled(metadata map[string]string) bool {
+	if _, ok := metadata[ErrorCodesFeatureMetadataKey]; ok {
+		return true
+	}
+	return false
+}
+
+// NewDaprError create a new DaprError using the supplied metadata and ErrorOptions
+// **Note**: As this code is in `Feature Preview`, it will only continue processing
+// if the ErrorCodes is enabled
+func NewDaprError(err error, metadata map[string]string, options ...ErrorOption) *DaprError {
+	// The following condition can be removed once the
+	// Error Codes Feature is GA
+	if !featureEnabled(metadata) {
+		return nil
+	}
+
+	if err == nil {
+		return nil
+	}
+
+	// Use default values
+	de := &DaprError{
+		err:            err,
+		description:    err.Error(),
+		reason:         NoReason,
+		httpCode:       503,
+		grpcStatusCode: codes.Internal,
+	}
+
+	// Now apply any requested options
+	// to overr
+	for _, option := range options {
+		option(de)
+	}
+
+	return de
+}
+
+// Error implements the error interface.
+func (c *DaprError) Error() string {
+	if c.err != nil {
+		return c.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (c *DaprError) Unwrap() error {
+	return c.err
+}
+
+// Description returns the description of the error.
+func (c *DaprError) Description() string {
+	if c.description != "" {
+		return c.description
+	}
+	return c.err.Error()
+}
+
+func (c *DaprError) SetDescription(description string) {
+	c.description = description
+}
+
+func (c *DaprError) SetResourceInfo(resourceInfo *ResourceInfo) {
+	c.resourceInfo = resourceInfo
+}
+
+func (c *DaprError) SetMetadata(md map[string]string) {
+	c.metadata = md
+}
+
+func WithResourceInfo(resourceInfo *ResourceInfo) ErrorOption {
+	f := func(de *DaprError) {
+		de.resourceInfo = resourceInfo
+	}
+	return f
+}
+
+func WithDescription(description string) ErrorOption {
+	f := func(de *DaprError) {
+		de.description = description
+	}
+	return f
+}
+
+func WithMetadata(md map[string]string) ErrorOption {
+	f := func(de *DaprError) {
+		de.metadata = md
+	}
+	return f
+}
```

Comment:
What's the use case for exposing these funcs? I don't think we should let these fields be mutable after creation.

Reply:
The `Set*` functions have been removed. The `With*` are used as optional at the time of creation.